src/libraries/Native/Unix/Common/pal_atomic.h

```diff
@@ -9,11 +9,14 @@
 #include "windows.h"
 #endif
 
+// The comparand should be the address to the pointer containing the value to compare to.
+// this matches Unix __atomic_compare_exchange_n, which compares the contents of the pointer.
 static int pal_atomic_cas_ptr(void* volatile* dest, void* exchange, void* comparand)
 {
 #if defined(TARGET_UNIX)
-    return __atomic_compare_exchange_n(dest, exchange, comparand, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+    return __atomic_compare_exchange_n(dest, comparand, exchange, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
 #elif defined(TARGET_WINDOWS)
-    return InterlockedCompareExchangePointer(dest, exchange, comparand) == comparand;
+    void *value = *((void**)comparand);
```

Comment:
This looks odd. I think the meaning of `pal_atomic_cas_ptr` arguments should match  `InterlockedCompareExchangePointer` .

Reply:
The problem here is that the linux API compares the contents of the ptr and the windows `InterlockedCompareExchangePointer` compares the values. So when calling this API in pal_collation.c we need to do:

```c
UCollator* pNull = NULL;

pal_atomic_cas_ptr(..., ..., &pNull);
```

by passing the address of pNull so that the linux implementation actually compares against NULL. We would need to #ifdef the callsite. Do you have a suggestion how to improve that?

Reply:
You can take the address of the argument in the Unix version of the helper.

Reply:
right, but then I would need to #ifdef the callsite no?

Reply:
No. The callsite would always use the `InterlockedCompareExchangePointer` style of arguments.

Reply:
> You can take the address of the argument in the Unix version of the helper.

Yeah I see what you're saying. I think I tried that, but since we would be passing `pNull` which is a `NULL` ptr, can we actually get the address of that?

Reply:
Yes, you should be able to get address of a variable holding null just fine.

Reply:
Ok, I'll try it out.

Reply:
looks I am the one confused Santi with this :-)

---
In reply to: [405146796](https://github.com/dotnet/runtime/pull/34645#discussion_r405146796) [](ancestors = 405146796)